### ISSUE TITLE
AVX512 code will need to change to handle the whole AVX10 business

### ISSUE THREAD

**Issue by workingjubilee** [dalek-cryptography/curve25519-dalek#693]
Hello.

You heard about the whole AVX10 business, I presume? With AVX10/256 and AVX10/512?

We will likely need to adjust what the "avx512" features even mean, unfortunately, as a result of  this. Fortunately, they're unstable, so we can do that! So it might need some modifications in dalek, but it's not like there will be any unforeseen complications from changing nightly features in code that didn't enable it intentionally... right?

https://github.com/dalek-cryptography/curve25519-dalek/blob/0964f800ab2114a862543ca000291a6e3531c203/curve25519-dalek/build.rs#L38-L44

...right?

**Comment by RalfJung** [comment]
In other words: building this crate may fail on nightly when the work on redoing these features happens. The compiler team considers automatic use of nightly features in crates that otherwise work on stable to be a breach of the nightly compiler contract, so this is not considered a breaking change -- but if this happens, it will mean that everyone using nightly to build this crate (or any crate that depends on this crate) will get compilation failures.

For context:
- https://github.com/rust-lang/rust/issues/120804
- https://github.com/rust-lang/rust/pull/124339
- https://github.com/SergioBenitez/version_check/pull/23

The preferred way to avoid this issue is to not enable nightly features automatically, but instead make them cargo features that need to be opt-in. That makes it possible to build the crate on nightly in exactly the same way that it would build on stable, which is extremely valuable for doing compiler development and testing.

Another option is to use something like [autocfg](https://docs.rs/autocfg/1.3.0/autocfg/struct.AutoCfg.html#method.probe_raw) to test whether the concrete kind of code you need works on the current compiler. If that code stops working, the crate can then automatically fall back to the stable code path. However, this reduces the usefulness of nightly as a testing vehicle: the point of nightly is that we can know that things will break long before they break on stable. To achieve this, however, it is important that testing a crate on nightly actually tests the thing that would happen when the nightly became stable. By making the same crate build different code on nightly vs stable, this is not the case any more.

**Comment by tarcieri** [comment]
I can't speak to auto-enabling nightly features, but in general I think `nightly` users need to expect that upstream rustc changes can break nightly builds at any time, and all we can do is release new versions of these crates when such upstream breaking changes happen, and nightly users need to upgrade to the latest versions which such breaking changes occur.

**Comment by RalfJung** [comment]
Of course nightly changes can break nightly crates (as in, crates that need nightly to build), but it should generally be the case that if a crate builds on stable, it also builds on nightly. This is crucial e.g. for narrowing down a stable-to-nightly regression.

**Comment by pinkforest** [comment]
Ref issues I raised earlier: https://github.com/dalek-cryptography/curve25519-dalek/issues/635

I've argued the AVX512 should be removed because it cannot be tested and maintained reliably.

Only consumer thing that benefits of it is 7950X now that intel slid up.

**Comment by workingjubilee** [comment]
Yes, I most certainly have an AVX512 CPU that can run the code in question, despite it being a desktop, consumer-oriented CPU, not even a workstation CPU, so I do not think you should disable it entirely, necessarily. :^)

And I am happy to help people run tests if necessary.

The main thing is that upcoming Intel CPUs will implement what Intel is calling "AVX10", which is specified here: https://cdrdv2-public.intel.com/784267/355989-intel-avx10-spec.pdf

Essentially, it allows new instruction patterns to be used while not requiring the full 512-bit zmm registers, but rather independently specifying that.

**Comment by workingjubilee** [comment]
Thus, we need to figure out some way of allowing you to say "I want to use this `AVX{version}` intrinsic, but I only need `{N}` bits."

**Comment by pinkforest** [comment]
How can we CI it? The original idea was that github used to have AVX512 capable runners at reasonable ratio and the nightly job tested it out. But now that GitHub has way less AVX512 capable runners testing it in CI has become impossible. Qemu has support but I don't know if it can be relied to a level to provide it as "Tier1 backend" or "Tier2" :thinking:

**Comment by tgross35** [comment]
I believe in stdarch we use the [Intel SDE](https://www.intel.com/content/www/us/en/developer/articles/tool/software-development-emulator.html), not sure how easy that is to get off the ground. @sayantn may know more.

**Comment by pinkforest** [comment]
PR's up:
- #695

**Comment by sayantn** [comment]
We are actually kind of stuck with SDE being our only option - `qemu-user` doesn't support AVX512, `qemu-system` does, but needs KVM support, which we don't have in the runners. SDE has some bugs too, when we tested with stdarch, v9.38 was quite buggy on linux (but surprisingly, the windows version was fine), so just be a little careful.

**Comment by sayantn** [comment]
I am still a bit confused here, what about the CPUs that are already released with AVX10.1-256? Does Intel not want to acknowledge their existence anymore, because they just removed the optional 512-bit FP/Int clause from everything

**Comment by nazar-pc** [comment]
> CPUs that are already released with AVX10.1-256?

Did they actually release any though?

**Comment by workingjubilee** [comment]
lol intel
